### PR TITLE
style: Share user agent cascade data across documents.

### PR DESCRIPTION
**Pull request by emilio**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18509)

### CODE REVIEW

**Comment by emilio** [comment]
This is on top of #18502. Closes https://bugzilla.mozilla.org/show_bug.cgi?id=1362532.

**Comment by SimonSapin** [comment]
Reviewed 11 of 11 files at r1, 1 of 1 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[components/style/stylesheet_set.rs, line 125 at r2](https://reviewable.io:443/reviews/servo/servo/18509#-Ku0hzQ-Rc1bLTyDD-Fv:-Ku0hzQ-Rc1bLTyDD-Fw:b-v0y9um) ([raw file](https://github.com/servo/servo/blob/061aac7c3a57e2ce1dcd54b869778062a710e088/components/style/stylesheet_set.rs#L125)):*
> ```Rust
>     // NB: Bound to the StylesheetSet lifetime when constructed, see
>     // StylesheetSet::flush.
>     collections: *mut PerOrigin<SheetCollection<S>>,
> ```

Less `unsafe`, nice :)

---

*[components/style/stylesheet_set.rs, line 175 at r3](https://reviewable.io:443/reviews/servo/servo/18509#-Ku0iEYKJzme-2Gr3F81:-Ku0iEYKJzme-2Gr3F82:bzh6pmg) ([raw file](https://github.com/servo/servo/blob/57540e4a6c8d1b780c55be243bfa7973f464d396/components/style/stylesheet_set.rs#L175)):*
> ```Rust
>         debug_assert_eq!(origin, Origin::UserAgent);
>         // Iterate over all the sheets to ensure state is consistent.
>         for (_, _) in self.origin_sheets(origin) {
> ```

What does "ensure state is consistent" mean? Does this iterator have side effects?

---

*[components/style/stylist.rs, line 198 at r3](https://reviewable.io:443/reviews/servo/servo/18509#-Ku0fkcxR9SUCAQ1Qsig:-Ku0fkcydSLM3pXCU56L:b-7581ei) ([raw file](https://github.com/servo/servo/blob/57540e4a6c8d1b780c55be243bfa7973f464d396/components/style/stylist.rs#L198)):*
> ```Rust
>     fn iter_origins_rev(&self) -> DocumentCascadeDataIter {
>         DocumentCascadeDataIter {
>             iter: self.per_origin.iter_origins(),
> ```

Should this be `iter_origins_rev`?

---

*[components/style/stylist.rs, line 302 at r3](https://reviewable.io:443/reviews/servo/servo/18509#-Ku0gqUbX5xbi4bTr7MN:-Ku0gqUcvkFR9XahiCcq:b-azgzi1) ([raw file](https://github.com/servo/servo/blob/57540e4a6c8d1b780c55be243bfa7973f464d396/components/style/stylist.rs#L302)):*
> ```Rust
>         self.author.add_size_of_children(ops, sizes);
> 
>         // FIXME(emilio): UA stuff is shared.
> ```

Please remove code rather than commenting it out, it’s in git history if we need it back. Or move it to an `impl MallocSizeOf for UserAgentCascadeDataCache`?

For the FIXME, file a bug and CC @nnethercote to measure `UA_CASCADE_DATA_CACHE` at the process level?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/18509)*

**Comment by emilio** [comment]
Review status: 10 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/style/stylesheet_set.rs, line 175 at r3](https://reviewable.io:443/reviews/servo/servo/18509#-Ku0iEYKJzme-2Gr3F81:-Ku0mV1ZTZLPg3oSESbG:bp4ss64) ([raw file](https://github.com/servo/servo/blob/57540e4a6c8d1b780c55be243bfa7973f464d396/components/style/stylesheet_set.rs#L175)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

What does "ensure state is consistent" mean? Does this iterator have side effects?
</blockquote></details>

Yes, it resets the `dirty` flag on the entries. But it's not a big deal because we flush them in a completely different way.

---

*[components/style/stylist.rs, line 198 at r3](https://reviewable.io:443/reviews/servo/servo/18509#-Ku0fkcxR9SUCAQ1Qsig:-Ku0mct0sMopW80qByW4:b-4kkiy6) ([raw file](https://github.com/servo/servo/blob/57540e4a6c8d1b780c55be243bfa7973f464d396/components/style/stylist.rs#L198)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Should this be `iter_origins_rev`?
</blockquote></details>

Nice catch :)

---

*[components/style/stylist.rs, line 302 at r3](https://reviewable.io:443/reviews/servo/servo/18509#-Ku0gqUbX5xbi4bTr7MN:-Ku0mbWN-mh2WyTuDGgY:bn4pwqp) ([raw file](https://github.com/servo/servo/blob/57540e4a6c8d1b780c55be243bfa7973f464d396/components/style/stylist.rs#L302)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Please remove code rather than commenting it out, it’s in git history if we need it back. Or move it to an `impl MallocSizeOf for UserAgentCascadeDataCache`?

For the FIXME, file a bug and CC @nnethercote to measure `UA_CASCADE_DATA_CACHE` at the process level?
</blockquote></details>

Done, and I will.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/18509)*

**Comment by SimonSapin** [comment]
@bors-servo r+

---

Reviewed 2 of 2 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/18509#-:-Ku0pisaBiO01BPMpTQd:bkg9nno)*

**Comment by emilio** [comment]
@bors-servo r-

 * Just want to make sure that the shutdown() patch makes debug builds happy.

**Comment by emilio** [comment]
Green: https://treeherder.mozilla.org/#/jobs?repo=try&revision=81ab3d06ac6c3e06da34472598a393f3abf6f8aa

@bors-servo r=SimonSapin

File: components/style/stylist.rs [hunk 1]
```diff
@@ -98,63 +247,64 @@ impl DocumentCascadeData {
...
+        self.user.add_size_of_children(ops, sizes);
+        self.author.add_size_of_children(ops, sizes);
 
-        for elem in self.precomputed_pseudo_element_decls.iter() {
-            if let Some(ref elem) = *elem {
-                sizes.mStylistPrecomputedPseudos += elem.shallow_size_of(ops);
-            }
-        }
+        // FIXME(emilio): UA_CASCADE_DATA_CACHE is shared, we should do whatever
+        // we do for RuleProcessorCache in Gecko.
```
Comment (heycam) [inline components/style/stylist.rs]: FWIW that is to assign all memory for nsCSSRuleProcessors in the RuleProcessorCache to `explicit/layout/rule-processor-cache`, and any non-cached nsCSSRuleProcessors are assigned to the style set.

**Comment by jdm** [comment]
```
error[E0277]: the trait bound `servo_arc::Arc<stylist::UserAgentCascadeData>: heapsize::HeapSizeOf` is not satisfied
   --> /home/servo/buildbot/slave/arm32/build/components/style/stylist.rs:158:38
    |
158 | #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
    |                                      ^^^^^^^^^^ the trait `heapsize::HeapSizeOf` is not implemented for `servo_arc::Arc<stylist::UserAgentCascadeData>`
    |
    = note: required by `heapsize::HeapSizeOf::heap_size_of_children`

   Compiling net_traits v0.0.1 (file:///home/servo/buildbot/slave/arm32/build/components/net_traits)
error: aborting due to previous error

```

**Pull request merged**